### PR TITLE
Ticket3889 in beam out beam components

### PR DESCRIPTION
**Pull request by John-Holt-Tessella**
### Description of work

Allow components to be moved in and out of the beam.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/3889

### Acceptance criteria

1. PV for enabled can be set to in and out and thus moves the component in and out of the beam
1. When out of the beam status is out of the beam and component is not included in the calculation
1. Beamline mode can be set through the GUI correctly and button changes value (needed monitor).
1. If the configuration contains an enable/disable parameter then an error is produced if there is not out of beam position on the associated driver,

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Has the author taken into account the multi-threaded nature of the code?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [ ] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?

### Functional Tests

- [ ] Do changes function as described? Add comments below that describe the tests performed.

### Final steps

- [ ] Reviewer has updated the submodule in the main EPICS repo? See **Reviewing work for the subModules of EPICS** in the [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by ISISBuildServer** [comment]
[34% (-0.01%) vs master 34%%20vs%20master%2034%25-red.svg)](http://epics-jenkins.isis.rl.ac.uk/job/inst_servers_pr_pipeline/job/PR-184/1/)

**Comment by ISISBuildServer** [comment]
[34% (+0.28%) vs master 34%%20vs%20master%2034%25-red.svg)](http://epics-jenkins.isis.rl.ac.uk/job/inst_servers_pr_pipeline/job/PR-184/2/)

File: ReflectometryServer/ioc_driver.py [hunk 1]
```diff
@@ -7,6 +7,9 @@
 
 logger = logging.getLogger(__name__)
 
+# Tolerance within which the position will be declared out of the beam. 1 mm tolerance.
+TOLERANCE_ON_OUT_OF_BEAM_POSITION = 1
```
Comment (Tom-Willemsen) [inline ReflectometryServer/ioc_driver.py]: Is this really a constant everywhere? Should this not be part of the configuration?

File: ReflectometryServer/ioc_driver.py [hunk 2]
```diff
@@ -23,6 +26,20 @@ def __init__(self, component, axis):
...
+
+    def is_for_component(self, component):
+        """
+        Does this driver use the component given.
+        Args:
+            component: the component to check
+
+        Returns: True if this ioc driver uses the component; false otherwise
+        """
+        return component == self._component
```
[the remark this answers is not in the document]
Comment (John-Holt-Tessella) [reply ReflectometryServer/ioc_driver.py]: yes.

File: ReflectometryServer/ioc_driver.py [hunk 3]
```diff
@@ -89,14 +109,31 @@ def _trigger_after_axis_value_change_listener(self, new_value, alarm_severity, a
...
+            distance_to_out_of_beam = abs(new_value - self._out_of_beam_position)
+            self._component.beam_path_rbv.is_in_beam = distance_to_out_of_beam > TOLERANCE_ON_OUT_OF_BEAM_POSITION
         self._component.beam_path_rbv.set_displacement(new_value, alarm_severity, alarm_status)
 
     def _get_set_point_position(self):
-        return self._component.beam_path_set_point.get_displacement()
+        if self._component.beam_path_set_point.is_in_beam:
+            displacement = self._component.beam_path_set_point.get_displacement()
+        else:
+            if self._out_of_beam_position is None:
```
[the remark this answers is not in the document]
Comment (John-Holt-Tessella) [reply ReflectometryServer/ioc_driver.py]: I prefer this. I want to indicate here that there are two conditions in and out of the beam.
If out of the beam then check for an error.

File: ReflectometryServer/ioc_driver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply ReflectometryServer/ioc_driver.py]: In that case use `is` rather than `==` to make it explicit.

**Comment by Tom-Willemsen** [comment]
Jenkins retest this please

**Comment by ISISBuildServer** [comment]
[34% (+0.21%) vs master 33%%20vs%20master%2033%25-red.svg)](http://epics-jenkins.isis.rl.ac.uk/job/inst_servers_pr_pipeline/job/PR-184/4/)

**Pull request merged**